be/src/storage/lake/transactions.cpp

```diff
@@ -51,42 +51,60 @@ StatusOr<TabletMetadataPtr> publish_version(TabletManager* tablet_mgr, int64_t t
+    // Do not delete txn logs if txns_size != 1, let gc do the work
+    // If the txn logs are deleted, it will be tricky to handle the situation of batch publish switching to single.
+
+    // for example:
+    // 1. the mode of publish is batch,
+    // 2. txn2 and txn3 have been published successfully and visible version in FE is updated to 3,
+    // 3. then txn4 and txn5 are published successfully in BE and the txn_log of txn4 and txn5 have been deleted,
+    // but FE do not get the response for some reason,
+    // 4. turn the mode of publish to single,
+    // 5. txn4 will be published in later publish task, but we can't judge what's the latest_version in BE and we can not reapply txn_log if
+    // txn logs have been deleted.
+    bool delete_txn_log = (txn_ids.size() == 1);
+    for (int i = 0; i < txn_ids.size(); i++) {
+        auto txn_id = txn_ids[i];
         auto log_path = tablet_mgr->txn_log_location(tablet_id, txn_id);
         auto txn_log_st = tablet_mgr->get_txn_log(log_path, false);
 
         if (txn_log_st.status().is_not_found()) {
-            return new_version_metadata_or_error(txn_log_st.status());
+            if (i == 0) {
+                // this may happen in two situations
+                // 1. duplicate publish in mode single
+                if (txn_ids.size() == 1) {
+                    return new_version_metadata_or_error(txn_log_st.status());
+                }
+
+                // 2. when converting from single publish to batch for txn log has been deleted,
+                // for example:
+                // the current mode of publish is single,
+                // txn2 has been published successfully and visible version in FE is updated to 2,
+                // then txn3 is published successfully in BE and the txn_log of txn3 has been deleted, but FE do not get the response for some reason,
+                // turn the mode of publish to batch,
+                // txn3 ,txn4, txn5 will be published in one publish batch task, so txn3 should be skipped just apply txn_log of txn4 and txn5.
+                auto missig_txn_log_meta = tablet_mgr->get_tablet_metadata(tablet_id, base_version + 1);
+                if (missig_txn_log_meta.status().is_not_found()) {
+                    // this should't happen
+                    LOG(WARNING) << "txn_log of txn: " << txn_id << " not found, and can not find the tablet_meta";
+                    return Status::InternalError("Both txn_log and corresponding tablet_meta missing");
+                } else if (!missig_txn_log_meta.status().ok()) {
+                    LOG(WARNING) << "txn_log of txn: " << txn_id << " not found, find the tablet_meta error: "
+                                 << missig_txn_log_meta.status().to_string();
+                    return new_version_metadata_or_error(missig_txn_log_meta.status());
+                } else {
+                    base_metadata = std::move(missig_txn_log_meta).value();
+                    break;
```

Comment:
I don't get it, why can you break the for loop here?

Reply:
Because the txn_log has been removed and the tablet_mata exsits, so we can skip the txn_log?

Reply:
But this is not the new version of the metadata to be generated.

Reply:
I get it， there should be continue